File: src-tauri/src/commands/structure.rs
use tauri::State;

use crate::models::{ColumnInfo, DatabaseInfo, DatabaseStructure, IndexInfo, SchemaInfo, TableInfo};
use crate::state::AppState;
use crate::storage::repositories::servers;

use super::connect_adapter;

#[tauri::command]
pub async fn list_databases(
    state: State<'_, AppState>,
    server_id: i64,
) -> Result<Vec<DatabaseInfo>, String> {
    // Metadata only (no keychain) to pick the database name to connect to.
    let server = servers::get_by_id_meta(&state.storage, server_id).map_err(|e| e.to_string())?;
    let db = server
        .default_database
        .as_deref()
        .unwrap_or_else(|| server.db_type.default_database_name());

    let adapter = connect_adapter(&state, server_id, db).await?;

    state
        .structure
        .list_databases(adapter)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn list_schemas(
    state: State<'_, AppState>,
    server_id: i64,
    database: String,
) -> Result<Vec<SchemaInfo>, String> {
    let adapter = connect_adapter(&state, server_id, &database).await?;

    state
        .structure
        .list_schemas(adapter)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn list_tables(
    state: State<'_, AppState>,
    server_id: i64,
    database: String,
    schema: String,
) -> Result<Vec<TableInfo>, String> {
    let adapter = connect_adapter(&state, server_id, &database).await?;

    state
        .structure
        .list_tables(adapter, &schema)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn list_columns(
    state: State<'_, AppState>,
    server_id: i64,
    database: String,
    schema: String,
    table: String,
) -> Result<Vec<ColumnInfo>, String> {
    let adapter = connect_adapter(&state, server_id, &database).await?;

    state
        .structure
        .list_columns(adapter, &schema, &table)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn list_indexes(
    state: State<'_, AppState>,
    server_id: i64,
    database: String,
    schema: String,
    table: String,
) -> Result<Vec<IndexInfo>, String> {
    let adapter = connect_adapter(&state, server_id, &database).await?;

    state
        .structure
        .list_indexes(adapter, &schema, &table)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn list_schemas_with_tables(
    state: State<'_, AppState>,
    server_id: i64,
    database: String,
) -> Result<DatabaseStructure, String> {
    let adapter = connect_adapter(&state, server_id, &database).await?;

    adapter
        .list_schemas_with_tables()
        .await
        .map_err(|e| e.to_string())
}

File: src/api/structure.ts
import { call } from './client';
import { RustCommand } from './commands';
import type {
  ColumnInfo,
  DatabaseInfo,
  DatabaseStructure,
  IndexInfo,
  SchemaInfo,
  TableInfo,
} from './types/structure.types';

export function listDatabases(serverId: number): Promise<DatabaseInfo[]> {
  return call<DatabaseInfo[]>(RustCommand.ListDatabases, { serverId });
}

export function listSchemas(
  serverId: number,
  database: string,
): Promise<SchemaInfo[]> {
  return call<SchemaInfo[]>(RustCommand.ListSchemas, { serverId, database });
}

export function listTables(
  serverId: number,
  database: string,
  schema: string,
): Promise<TableInfo[]> {
  return call<TableInfo[]>(RustCommand.ListTables, {
    serverId,
    database,
    schema,
  });
}

export function listColumns(
  serverId: number,
  database: string,
  schema: string,
  table: string,
): Promise<ColumnInfo[]> {
  return call<ColumnInfo[]>(RustCommand.ListColumns, {
    serverId,
    database,
    schema,
    table,
  });
}

export function listIndexes(
  serverId: number,
  database: string,
  schema: string,
  table: string,
): Promise<IndexInfo[]> {
  return call<IndexInfo[]>(RustCommand.ListIndexes, {
    serverId,
    database,
    schema,
    table,
  });
}

/** Schemas + tables in one call — ideal for the initial sidebar tree */
export function listSchemasWithTables(
  serverId: number,
  database: string,
): Promise<DatabaseStructure> {
  return call<DatabaseStructure>(RustCommand.ListSchemasWithTables, {
    serverId,
    database,
  });
}
